### PR TITLE
Org webhooks #1028

### PR DESCRIPTION
**Pull request by guptanubhav**
#1028 Implemented the Organization Webhook. Added the tests as well. Updated the urls of of repository hooks in some places. Added a ensure condition to check for hookid(added Please Review in comment. Please check if this should be there)

### CODE REVIEW

**Comment by Anubhav10** [comment]
6 integration tests I made are failing(All in OrganizationHooksClientTests.cs). I think this is because authenticated user has to be an admin of the organization being managed and OAuth tokens require the admin:org_hook scope.
Need some help, i can't figure out how to go about it.

**Comment by ryangribble** [comment]
Hey @Anubhav10 thanks heaps for the contribution!  Ive done a first pass review and made several comments... Hopefully you don't think Im being overly picky!  This is a fantastic contribution and in general you've nailed all the key elements - eg standard and observable implementations, great unit and integration tests, added any new files to all the csproj projects and so on.  

In terms of the feedback ive commented - most are around tidy up things like whitespace and I also highlighted a couple of things around consistency in naming/doc - eg http vs https on API doc links, or using ```org``` rather than ```organizationName``` for the parameters.   I realise that there are already some instances of inconsistencies within octokit in these areas but I guess my approach is when adding new work to try to conform with the "majority" consistency, and then hopefully we will also start to tackle the minority items that are inconsistent.  You also seem to have made a couple of inadvertent edits to the csproj files around nuget package version and app.config binding redirects etc.

To address a couple of specific things:

- Your comment about integration tests not running - it looks like you're setting "octokit" org in the test fixture, whereas you should be using a specific integration testing organization that you setup for yourself (accessed via the ```Helper``` class).  I made a more detailed comment in the location where this was being done

- With unit tests, when there are parameters that are being checked for empty strings as well as null (eg ```Ensure.ArgumentNotNullOrEmptyString(organizationName)````) you should add a unit test that verifies that.  So in your case, you've added the ```EnsuresNonNullArguments()``` unit test, but you could also add a ```EnsuresNonEmptyArguments()``` test that checks the exceptions are thrown when ```""``` are passed in for the various arguments.  If you search the code base, there should be some other examples of that if you need to see.

- Your [PleaseReview] comments about NotNull assertions on HookId are not required, as hookId being an ```int``` can never be ```null``` :grinning:

**Comment by Anubhav10** [comment]
Thanks for the review. I'll make these changes as soon as possible. It may take a few days as I have my mid-sem exams in the next week.

**Comment by ryangribble** [comment]
@shiftkey neither appveyor or travis builds seemed to fire on this one?  The fact this is a change about organization hooks and had the "octokit" organisation mentioned in integration tests, hopefully is just a co-incidence!  :grinning:

**Comment by shiftkey** [comment]
@ryangribble uh, that's, neat! Will have a look....

**Comment by shiftkey** [comment]
@ryangribble I was able to re-send the webhook to AppVeyor but the Travis integration doesn't have that option. Oh well, it should trigger when the PR is next synced...

**Comment by Anubhav10** [comment]
Sorry for the delay. I have made the changes
1. Changed organizationName to org
2. Added EnsuresNonEmptyArguments() in OrganizationHooksClientTest.cs
3. Fixed the problems in integrationTests I added for org webhooks after adding the test account
4. And other changes as asked by @ryangribble 
About the csproj and app.config changes, there was a problem in the test runner in my VS. It wasn't showing any tests. So, I updated all the Nuget Packages(It was a solution on stackoverflow), it still didn't work. I guess the update made those changes. There was a merge conflict when I commiting this time in the csproj, I have replaced these files with the one in master.

There are references to the documentation in https. Do you want me to change them all?

The integration tests have been skipped for OrganizationMemberClientTest, and the reasons given are "Admin/Member filter doesn't work unless the requester is an organization member" ,"TwoFactor filter can't be used unless the requester is an organization owner". These problems will come in org webhooks too, unless one is the admin of the org the tests will fail, so should I skip them?

One Of the integration test is failing. Test Name: Octokit.Tests.Integration.Clients.OrganizationHooksClientTests+TheCreateMethod.CreateAWebHookForTestOrganization.  The reason it is failing is that the 'active' value I send is false, but in the created hook active value is true. I guess this is a problem in the GitHub API.
Also the value of secret is returned as 'HMAC hex digest of the body, using the secret as the key' I guess( there are ******) instead of the secret, so I have commented the assert to check the value of config, and remove the secret for now. This test will pass if we change the active to true in assert check, but that won't be right!! :grinning:

**Comment by Anubhav10** [comment]
The travis-ci is passing this time!! :grinning:

File: Octokit.Reactive/Clients/IObservableOrganizationsClient.cs [hunk 1]
```diff
@@ -16,6 +16,12 @@ public interface IObservableOrganizationsClient
         IObservableTeamsClient Team { get; }
 
         /// <summary>
+        /// A client for GitHub's Organization Hooks API.
+        /// </summary>
+        /// <remarks>See <a href="http://developer.github.com/v3/orgs/hooks/">Hooks API documentation</a> for more information.</remarks>
+        IObservableOrganizationHooksClient Hooks { get; }
```
Comment (M-Zuber) [inline Octokit.Reactive/Clients/IObservableOrganizationsClient.cs]: This should be in the singular - `Hook` to be consistent with the other sub-client properties

File: Octokit.Reactive/Clients/ObservableOrganizationsClient.cs [hunk 2]
```diff
@@ -35,6 +36,12 @@ public ObservableOrganizationsClient(IGitHubClient client)
         public IObservableTeamsClient Team { get; private set; }
 
         /// <summary>
+        /// A client for GitHub's Organization Hooks API.
+        /// </summary>
+        /// <remarks>See <a href="http://developer.github.com/v3/orgs/hooks/">Hooks API documentation</a> for more information.</remarks>
+        public IObservableOrganizationHooksClient Hooks { get; private set; }
```
Comment (M-Zuber) [inline Octokit.Reactive/Clients/ObservableOrganizationsClient.cs]: This should be in the singular - Hook to be consistent with the other sub-client properties (will flow from a refactor of the interface but figured I'd mention it here also)

File: Octokit.Tests.Integration/Clients/OrganizationHooksClientTests.cs [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+                Assert.Equal(baseHookUrl, hook.Url);
+                Assert.Equal(baseHookUrl + "/pings", hook.PingUrl);
+                Assert.NotNull(hook.CreatedAt);
+                Assert.NotNull(hook.UpdatedAt);
+                Assert.Equal(webHookConfig.Keys, hook.Config.Keys);
+                //Assert.Equal(webHookConfig.Values, hook.Config.Values);
+                Assert.Equal(false, hook.Active);
+            }
+
+            Dictionary<string, string> CreateExpectedConfigDictionary(Dictionary<string, string> config, string url, OrgWebHookContentType contentType)
```
Comment (M-Zuber) [inline Octokit.Tests.Integration/Clients/OrganizationHooksClientTests.cs]: This method does not seem to be doing anything. Could you please explain it's purpose?

File: Octokit.Tests.Integration/fixtures/OrganizationsHooksCollection.cs [hunk 4]
```diff
@@ -0,0 +1,10 @@
+﻿using Xunit;
+
+namespace Octokit.Tests.Integration.fixtures
```
Comment (M-Zuber) [inline Octokit.Tests.Integration/fixtures/OrganizationsHooksCollection.cs]: Namespace parts should be uppercased

File: Octokit.Reactive/Clients/IObservableOrganizationsClient.cs (hunk 1 above)
Comment (Anubhav10) [inline Octokit.Reactive/Clients/IObservableOrganizationsClient.cs]: I added these properties based on the iobservablerepositoriesclient, it was 'Hooks' there. I thought that is how it should be named here too. I'll change it if you want

File: Octokit/Clients/IOrganizationsClient.cs [hunk 5]
```diff
@@ -25,6 +25,12 @@ public interface IOrganizationsClient
         ITeamsClient Team { get; }
 
         /// <summary>
+        /// A client for GitHub's Organization Hooks API.
+        /// </summary>
+        /// <remarks>See <a href="https://developer.github.com/v3/orgs/hooks/">Hooks API documentation</a> for more information.</remarks>
+        IOrganizationHooksClient Hooks { get; }
```
Comment (M-Zuber) [inline Octokit/Clients/IOrganizationsClient.cs]: This should be in the singular - Hook to be consistent with the other sub-client properties

File: Octokit.Tests.Integration/Clients/OrganizationHooksClientTests.cs (hunk 3 above)
Comment (Anubhav10) [inline Octokit.Tests.Integration/Clients/OrganizationHooksClientTests.cs]: It creates a expected config, which is then compared with the returned config in the response

File: Octokit.Reactive/Clients/IObservableOrganizationsClient.cs (hunk 1 above)
Comment (M-Zuber) [inline Octokit.Reactive/Clients/IObservableOrganizationsClient.cs]: @ryangribble is the master of consistency around here, but I believe that (in a separate  PR) IObservableRepositoriesClient (and probably IRepositoriesClient) would need to be changed also

File: Octokit.Tests.Integration/fixtures/OrganizationsHooksCollection.cs (hunk 4 above)
Comment (Anubhav10) [inline Octokit.Tests.Integration/fixtures/OrganizationsHooksCollection.cs]: The folder 'fixtures' was also created. I had to use the lowercase.

File: Octokit.Reactive/Clients/ObservableOrganizationHooksClient.cs [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        {
+            Ensure.ArgumentNotNullOrEmptyString(org, "org");
+
+            return _client.Ping(org, hookId).ToObservable();
+        }
+
+        /// <summary>
+        /// Deletes a hook for a organization
+        /// </summary>
+        /// <param name="owner"></param>
```
Comment (M-Zuber) [inline Octokit.Reactive/Clients/ObservableOrganizationHooksClient.cs]: There is no corresponding parameter for this, so the Appeyor build is failing

Comment (Anubhav10) [inline Octokit.Reactive/Clients/ObservableOrganizationHooksClient.cs]: I'll make this change! Thanks!

File: Octokit.Tests.Integration/Clients/OrganizationHooksClientTests.cs (hunk 3 above)
Comment (M-Zuber) [inline Octokit.Tests.Integration/Clients/OrganizationHooksClientTests.cs]: But as far as I can see the method is just returning the config dictionary passed in...
(after `Union` with an empty dictionary and then `ToDictionary` for some reason)

File: Octokit.Tests.Integration/fixtures/OrganizationsHooksCollection.cs (hunk 4 above)
Comment (M-Zuber) [inline Octokit.Tests.Integration/fixtures/OrganizationsHooksCollection.cs]: Can you rename the folder to uppercase?

File: Octokit.Reactive/Clients/IObservableOrganizationsClient.cs (hunk 1 above)
Comment (ryangribble) [inline Octokit.Reactive/Clients/IObservableOrganizationsClient.cs]: Yep, there has been a move towards ensuring all property names for clients are singular (#53 #1035 #1036 #375 #1038), 
so I agree these should be `IOrganizationHooksClient Hook { get; private set; }`

It also makes me realise we have quite a few other clients that still need to get resolved!

File: Octokit.Tests.Integration/fixtures/OrganizationsHooksCollection.cs (hunk 4 above)
Comment (ryangribble) [inline Octokit.Tests.Integration/fixtures/OrganizationsHooksCollection.cs]: There are already 2 other existing classes in here with that lowercase name.  I agree it doesnt look nice but from a consistency point of view, this PR is consistent with all other occurences.  A separate issue/PR can be raised to tidy this up

Comment (M-Zuber) [inline Octokit.Tests.Integration/fixtures/OrganizationsHooksCollection.cs]: Will do.
being consistently consistent can be hard :)

File: Octokit.Tests.Integration/fixtures/OrganizationsHooksFixture.cs [hunk 7]
```diff
@@ -0,0 +1,42 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace Octokit.Tests.Integration.fixtures
+{
+    public class OrganizationsHooksFixture : IDisposable
+    {
+        readonly IGitHubClient _github;
+        readonly OrganizationHook _hook;
+        private string organizationFixture;
```
Comment (ryangribble) [inline Octokit.Tests.Integration/fixtures/OrganizationsHooksFixture.cs]: local class variable should be named `_organizationFixture` and be `readonly` if possible

File: Octokit.Reactive/Clients/IObservableOrganizationsClient.cs (hunk 1 above)
Comment (M-Zuber) [inline Octokit.Reactive/Clients/IObservableOrganizationsClient.cs]: Opened #1207 to go into this a bit more

File: Octokit.Tests.Integration/fixtures/OrganizationsHooksFixture.cs [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+        public string org { get { return organizationFixture; } }
+
+        public OrganizationHook ExpectedHook { get { return _hook; } }
+
+        public void Dispose()
+        {
+            _github.Organization.Hooks.Delete(organizationFixture,_hook.Id);
+        }
+
+        static OrganizationHook CreateHook(IGitHubClient github, string _organizationFixture)
```
Comment (ryangribble) [inline Octokit.Tests.Integration/fixtures/OrganizationsHooksFixture.cs]: method variable shouldn't be named with underscore

File: Octokit.Tests.Integration/fixtures/OrganizationsHooksCollection.cs (hunk 4 above)
Comment (Anubhav10) [inline Octokit.Tests.Integration/fixtures/OrganizationsHooksCollection.cs]: I already changed the name on the other 2 classes as well should I undo the changes?

Comment (M-Zuber) [inline Octokit.Tests.Integration/fixtures/OrganizationsHooksCollection.cs]: done - #1208

File: Octokit.Tests.Integration/fixtures/OrganizationsHooksFixture.cs (hunk 8 above)
Comment (Anubhav10) [inline Octokit.Tests.Integration/fixtures/OrganizationsHooksFixture.cs]: Then the same problem would be there in RepositoriesHooksFixture

File: Octokit.Tests.Integration/fixtures/OrganizationsHooksCollection.cs (hunk 4 above)
Comment (Anubhav10) [inline Octokit.Tests.Integration/fixtures/OrganizationsHooksCollection.cs]: I'll undo the changes in the 2 classes and make it Fixtures in the classes I am adding here

File: Octokit.Tests.Integration/fixtures/OrganizationsHooksFixture.cs (hunk 8 above)
Comment (ryangribble) [inline Octokit.Tests.Integration/fixtures/OrganizationsHooksFixture.cs]: I had a look at [RepositoriesHooksFixture](https://github.com/octokit/octokit.net/blob/0e664c23aeb21d25a96c8acf50016c9c5b56b1ce/Octokit.Tests.Integration/fixtures/RepositoriesHooksFixture.cs) but it appears to be "correct" ie class variables are readonly and named with underscores, while method variables do not have underscores.

Must've just been something that happened inadvertently on your side after you copied that class and modified it for Org hooks?

File: Octokit.Tests.Integration/fixtures/OrganizationsHooksCollection.cs (hunk 4 above)
Comment (ryangribble) [inline Octokit.Tests.Integration/fixtures/OrganizationsHooksCollection.cs]: > I already changed the name on the other 2 classes as well should I undo the changes?
Whatever you prefer.  I was trying to avoid you needing to tackle more problems on this PR that were already existing (I also wasnt sure what the repercussions of renaming that folder and or namespace might be) which is why I thought a separate PR gets it off your plate.  But if you've done it then no big deal!

File: Octokit.Tests.Integration/Clients/OrganizationHooksClientTests.cs (hunk 3 above)
Comment (ryangribble) [inline Octokit.Tests.Integration/Clients/OrganizationHooksClientTests.cs]: Agree... also it seems like it isnt doing anything with the `url` and `contentType` parameters passed in

File: Octokit.Reactive/Clients/IObservableOrganizationsClient.cs (hunk 1 above)
Comment (Anubhav10) [inline Octokit.Reactive/Clients/IObservableOrganizationsClient.cs]: done

File: Octokit.Tests.Integration/fixtures/OrganizationsHooksCollection.cs (hunk 4 above)
Comment (Anubhav10) [inline Octokit.Tests.Integration/fixtures/OrganizationsHooksCollection.cs]: done

File: Octokit.Tests.Integration/fixtures/OrganizationsHooksFixture.cs (hunk 7 above)
Comment (Anubhav10) [inline Octokit.Tests.Integration/fixtures/OrganizationsHooksFixture.cs]: done

File: Octokit.Tests.Integration/fixtures/OrganizationsHooksFixture.cs (hunk 8 above)
Comment (Anubhav10) [inline Octokit.Tests.Integration/fixtures/OrganizationsHooksFixture.cs]: The integration tests weren't working, and the error was something related to fixtures, hence I tried out some things maybe then it got changed.

**Comment by Norbo11** [comment]
Did the organization hooks client get removed at any point, or did it never make it into Octokit? It appears that GitHubClient.Organization.Hooks no longer exists...

**Comment by ryangribble** [comment]
Hi @Norbo11 this PR was not completed and thus hasn't been merged

**Comment by AraHaan** [comment]
So this will allow me to see if GitHub posts to a webhook I want to GitHub application to have to do specific stuff (like checking if the issue that posted to it has a specific label, is a pull request, edits specific files) and posts in it's status/checks?

Currently I got a untested and unrunnable ruby CI that I am not sure will work and I am moving it to .NET Core to C# that I know a lot more on.

**Comment by shiftkey** [comment]
Assigning myself to see if we can rebase this on top of `master` and address those conflicts

**Comment by shiftkey** [comment]
I also see #1884 has been opened and might be closer to merging

**Comment by shiftkey** [comment]
After a quick glance at #1884, I'm going to close this PR out in favour of the newer as it seems ready to review.
